### ISSUE TITLE
Neutrino 9 beta/release candidate feedback

### ISSUE THREAD

**Issue by edmorley** [neutrinojs/neutrino#1129]
Hi!

We've just published a beta release of Neutrino 9 and are keen to have people trying it out/giving feedback. Notable changes:
* now using webpack 4, Babel 7, ESLint 6, and latest versions of test runners.
* significant performance improvements ([see here](https://github.com/neutrinojs/neutrino/issues?q=label%3Aperformance)).
* switched from using webpack/ESLint/test runner's APIs to using their native CLIs - which fixes a number of bugs, and means that the full set of CLI features are available for each, not just what we've emulated via the API.
* the `neutrino --inspect` feature now produces a fully-working webpack config, for easier debugging.

Full changelogs:
* [v8.3.0 -> v9.0.0-beta.0](https://github.com/neutrinojs/neutrino/compare/v8.3.0...d48eb35)
* [v9.0.0-beta.0 -> v9.0.0-beta.1](https://github.com/neutrinojs/neutrino/compare/d48eb35...1c082ac)
* [v9.0.0-beta.1 -> v9.0.0-rc.0](https://github.com/neutrinojs/neutrino/compare/1c082ac...e9c21b8)
* [v9.0.0-rc.0 -> v9.0.0-rc.1](https://github.com/neutrinojs/neutrino/compare/e9c21b8...v9.0.0-rc.1)
* [v9.0.0-rc.1 -> v9.0.0-rc.2](https://github.com/neutrinojs/neutrino/compare/v9.0.0-rc.1...v9.0.0-rc.2)
* [v9.0.0-rc.2 -> v9.0.0-rc.3](https://github.com/neutrinojs/neutrino/compare/v9.0.0-rc.2...v9.0.0-rc.3)
* [v9.0.0-rc.3 -> v9.0.0-rc.4](https://github.com/neutrinojs/neutrino/compare/v9.0.0-rc.3...v9.0.0-rc.4)
* [v9.0.0-rc.4 -> v9.0.0-rc.5](https://github.com/neutrinojs/neutrino/compare/v9.0.0-rc.4...v9.0.0-rc.5)
* [v9.0.0-rc.5 -> v9.0.0](https://github.com/neutrinojs/neutrino/compare/v9.0.0-rc.5...v9.0.0)

The docs for the new version can be found here:
~https://master.neutrinojs.org~ https://neutrinojs.org/

**Note that during the beta phase, you'll need to explicitly install the `@next` version of any Neutrino monorepo packages when following the docs, otherwise you will install the latest stable (v8.3.0) packages instead.**

For example, to use v9's `create-project`, run:
```bash
# NPM
npx @neutrinojs/create-project@next <directory-name>

# Yarn
# (There's a bug with `yarn create` and using a version number, so we run the commands
#  manually. Make sure the directory returned by `yarn global bin` is on `PATH` first.)
yarn global add @neutrinojs/create-project@next
create-project <directory-name>
```

Or to upgrade an existing project, modify the `neutrino` and `@neutrinojs/*` entries in your `package.json` to use version ~`^9.0.0-rc.5`~ `^9.0.0` and then see:
~https://master.neutrinojs.org/migration-guide/~ https://neutrinojs.org/migration-guide/

If you find any bugs (including things in the migration guide/docs that aren't clear enough) please comment here or file issues, so we can sort out any rough edges :-)

Many thanks!

**Comment by edmorley** [comment]
To give an example for the performance improvements - I created a project using Neutrino's `create-project` with both Neutrino 8 and Neutrino 9, that used `@neutrinojs/airbnb` + `@neutrino/react` + `@neutrino/jest` and timed various workflows:

|    Command   |        Neutrino 8       |    Neutrino 9 beta    |
|:------------:|:-----------------------:|:---------------------:|
| `yarn build` | Cold: 12.8s <br/> Warm: 12.4s | Cold: 7.0s <br/> Warm: 4.2s |
| `yarn start` |  Cold: 8.0s <br/> Warm: 7.4s  | Cold: 6.2s <br/> Warm: 5.4s |
|  `yarn test` |  Cold: 4.8s <br/> Warm: 4.4s  | Cold: 3.8s <br/> Warm: 3.0s |
|  `yarn lint` |  Cold: 5.3s <br/> Warm: 5.3s  | Cold: 2.5s <br/> Warm: 2.0s |

(`yarn start` was timed by forcing a `process.exit()` after the "compilation complete" message)

The speedup is even greater on projects that are larger than the `create-project` "hello world".

Please do let us know how much of a difference upgrading makes! :-)

**Comment by Shyiy** [comment]
After upgrading to 9 beta from 8 and doing various minor adjustments following the migration guide, everything seems to work nicely by me except one thing.

I was using the [webpack-pwa-manifest](https://github.com/arthurbergmz/webpack-pwa-manifest) plugin to generate and inject a PWA manifest in the html. I verified that the plugin is supposed to be updated for Webpack v4.

I add the plugin to the end of the `use` config option in `.neutrinorc.js` on production with the following entry:

```
neutrino =>
    neutrino.config.plugin("webpack-pwa-manifest").use(require.resolve("webpack-pwa-manifest"), [
      {
        name: "appTitle",
        short_name: "appShortName",
        description: "appDescription",
        [... rest of the plugin config]
      }
    ])
```

The plugin is correctly picked up and executed, generating all its files and manifest in the build folder, it however fails to inject its contents in the `index.html` file. I've investigated a bit and by running ` neutrino --inspect --mode production` it looks like the PWA manifest plugin is correctly loaded after the needed `html-webpack-plugin` plugin. When adding some debugging statements to the PWA manifest plugin it seems however that the necessary webpack compilation hook that is supposed to be provided by `html-webpack-plugin`, `htmlWebpackPluginBeforeHtmlProcessing`, is not present. Actually none of the hooks advertised in the [HtmlWebpackPlugin docs](https://github.com/jantimon/html-webpack-plugin#events) seem to be present.

Is this a possible bug introduced by the current neutrino v9 beta? Note that the same config/code worked correctly before.

**Comment by edmorley** [comment]
Hi! Thank you for trying out the beta.

For Neutrino 9, we've updated `html-webpack-plugin` to 4.x since 3.x has issues with webpack 4 (particularly when using multiple entrypoints):
https://github.com/neutrinojs/neutrino/blob/7e5e3abb0bed6fca5f87679ff0a0ec0d276581d1/packages/html-template/package.json#L26

The 4.x release of `html-webpack-plugin` adjusted the hook names, so plugins that use them will need to be updated. The reason those new hook names don't appear in the `html-webpack-plugin` readme, is that the new changes haven't yet been merged to `master`, and are currently only on the `webpack-4` branch / PR:
https://github.com/jantimon/html-webpack-plugin/tree/webpack-4
https://github.com/jantimon/html-webpack-plugin/pull/953

Perhaps it might be worth filing an issue over at https://github.com/arthurbergmz/webpack-pwa-manifest to see if they can add support even before the final `html-webpack-plugin@4.0.0` release? :-)

**Comment by Shyiy** [comment]
You're absolutely right, thanks for the in depth answer, I got confused by the master/beta README of the newly used plugin.

Nothing left for me to nag about then, it all works nicely :) 

Only a small typo that I noticed in the migration guide:
```
The minimum supported Node.js version is 8.10. Node.js 9 is no longer supported #792. (should be 6)
```

**Comment by edmorley** [comment]
@mattmcdonald-uk hi! Does the output from `yarn neutrino --inspect --mode development` shed any light? (If upgrading from Neutrino 8.3.0, I'd also compare it with the output from `yarn neutrino start --inspect-new`).

@Shyiy - great! Thank you for trying it out and the feedback. Re the docs Node.js version, that part if definitely unclear, it should probably say "Node.js 6 and 9 are no longer supported".

**Comment by mattmcdonald-uk** [comment]
@edmorley If I turn `extract` off it works fine, with `extract` on it doesn't.

The different this makes to the development config output is removing this:
```
 /* neutrino.config.module.rule('style').use('extract') */
          {
            loader: '/Users/matt/Dropbox/Freelance/MyVitae/Code/app/node_modules/mini-css-extract-plugin/dist/loader.js'
          },
```

**Comment by edmorley** [comment]
Ah so `extract` isn't supposed to be enabled in development builds. What is the `mode` set to in the inspect output? Is `--mode development` being passed to webpack-dev-server? (or to webpack, if not using the dev server)

**Comment by mattmcdonald-uk** [comment]
It's being started with `webpack-dev-server --mode development`

**Comment by edmorley** [comment]
Ah - are you setting custom `extract` options?

Currently the web preset disables extract when not in production, here:
https://github.com/neutrinojs/neutrino/blob/7e5e3abb0bed6fca5f87679ff0a0ec0d276581d1/packages/web/index.js#L32

However by passing custom options, that will be overwritten, and extract always enabled:
https://github.com/neutrinojs/neutrino/blob/7e5e3abb0bed6fca5f87679ff0a0ec0d276581d1/packages/style-loader/index.js#L21-L27
https://github.com/neutrinojs/neutrino/blob/7e5e3abb0bed6fca5f87679ff0a0ec0d276581d1/packages/style-loader/index.js#L53

I believe the current implementation is so that users can force extraction to be enabled always (and not only in production) if desired. However that results in counter-intuitive behaviour in this case.

I think we should either:
1. Move the production conditional from the web preset into `@neutrinojs/style-loader` instead, so consumers of the web preset don't have to remember to conditionally pass any extract options. (But now they can't force it to be enabled in development)
2. Separate out the extraction on/off from the options. eg add an `enabled` key under `extract` or similar.

@mattmcdonald-uk @eliperelman, thoughts?

**Comment by mattmcdonald-uk** [comment]
@edmorley With my original issue I hadn't set the `extract` option, I just discovered it as a workaround for my problem.

From what I can see, currently `extract` is being used in development mode if left unset in your config.

**Comment by edmorley** [comment]
I've verified locally that this doesn't happen when using the web preset - are you using `@neutrinojs/style-loader` directly? Could you paste your `.neutrinorc.js`? (I really should have asked for this at the start hehe)

**Comment by mattmcdonald-uk** [comment]
Sure, this is what I'm using:
```
const Dotenv = require('dotenv-webpack')

module.exports = {
    use: [
        [
            '@neutrinojs/react', {
                html: {
                    title: 'Project',
                },
                devServer: {
                    port: 8080,
                },
                publicPath: '/',
            },
        ],
        '@neutrinojs/style-loader',
        (neutrino, envs = []) => {
            neutrino.config.plugin('dotenv').use(Dotenv, envs)
        },
    ],
}
```

**Comment by edmorley** [comment]
Ah right that explains it :-) The `@neutrinojs/react` preset includes `@neutrinojs/style-loader` inside it, so `@neutrinojs/style-loader` should not be included separately. Perhaps we need to make that clearer in the docs, or add a check inside `@neutrinojs/style-loader` to make sure it's not being included twice?

**Comment by mattmcdonald-uk** [comment]
Ah. That would explain it!

So how would I configure `style-loader`? I'd gone back to a basic config to try and get it working, but want to be able to use `sass-loader` too. This was is what I would want to achieve:

```
'@neutrinojs/style-loader', {
                test: /\.(css|sass|scss)$/,
                loaders: [{
                    loader: 'sass-loader',
                    useId: 'sass',
                }],
            },
```

**Comment by edmorley** [comment]
The `@neutrinojs/react` preset has a `style` option, that controls what's passed to `@neutrinojs/style-loader`:
https://master.neutrinojs.org/packages/web/#preset-options
https://github.com/neutrinojs/neutrino/blob/7e5e3abb0bed6fca5f87679ff0a0ec0d276581d1/packages/web/index.js#L30-L33
https://github.com/neutrinojs/neutrino/blob/7e5e3abb0bed6fca5f87679ff0a0ec0d276581d1/packages/web/index.js#L176

This isn't the first time that those options have been hard for people to find -- so we should probably come up with a way to adjust the docs to make it clearer. Any ideas? :-)

**Comment by mattmcdonald-uk** [comment]
Great, everything's working now. Thanks for helping me sort that (and for releasing v9 - hot loading is working properly for me again! 🎉)

I initially took a guess that the key might be `styleLoader` or `'style-loader'`, I don't know if there's any benefit to making the key more directly related to the name of the thing you're configuring?

Other than that I think a mention that the `style-loader` middleware is already included with the react (and any other) preset in the docs, along with a note on the key to use would be handy. Probably a note on the key to use in the `react` section would be helpful too.

An additional CLI warning that you shouldn't include `style-loader` directly when you're using the react preset would probably be useful too if you had missed the warnings within the docs.

**Comment by ratson** [comment]
@edmorley I see "re-license at MIT" was removed from  https://github.com/neutrinojs/neutrino/issues/693 title, and the beta release is still MPL-2.0.
Is that not going to happen for v9 release? If `webpack-chain` would follow the same?

**Comment by edmorley** [comment]
@mattmcdonald-uk, glad it's working now! I'll have a think about what changes I can make.

@ratson, yeah I discussed with Eli on IRC and said that (a) the two changes were unrelated (and both significant) so shouldn't be bundled into the same PR, (b) we needed to figure out what permissions we did/didn't need from previous contributors before we relicence. Is lack of MIT licence blocking your usage of Neutrino?

**Comment by ratson** [comment]
Not currently, just wondering if releasing with a more permissive license is still on the plan.

**Comment by edmorley** [comment]
Here are two PRs for projects upgrading from Neutrino 8 to 9:
* mozilla-frontend-infra/firefox-health-dashboard#158
* mozilla-frontend-infra/js-perf-dashboard#86

Some of the issues encountered for which we could improve the UX:
* usages of the `html.links` option had to be replaced (but was silently ignored, meaning it was missed during the first pass of the migration)
* the `./static/` directory was no longer copied since we removed that feature (we do mention this in the release notes, but perhaps we can do a directory exists check combined with seeing if they've added `@neutrinojs/copy` and if not, warn?)
* at one point in the PR iterations, `--optimize-minimize` was being passed to `webpack` rather than `--mode` (something else wasn't working, so various permutations of settings were being tried). There's not much we can/should do about this IMO, but I've filed webpack/webpack-cli#624 for making it show a deprecation warning.)
* not specific to the migration, but both the `@neutrinojs/react` and `@neutrinojs/web` presets were being used simultaneously, causing confusion due to the web preset clobbering the settings from the react preset. Perhaps we should have the top-level presets error out in this case?
* there were lots of stale dependencies in the yarn lockfiles - we should perhaps emphasise the need to regenerate the lockfile

**Comment by edmorley** [comment]
The following have been addressed in #1160.

> Only a small typo that I noticed in the migration guide:
> 
> ```
> The minimum supported Node.js version is 8.10. Node.js 9 is no longer supported #792. (should be 6)
> ```


> * usages of the `html.links` option had to be replaced (but was silently ignored, meaning it was missed during the first pass of the migration)
> * there were lots of stale dependencies in the yarn lockfiles - we should perhaps emphasise the need to regenerate the lockfile

I'll open a separate PR for adding checks to prevent presets from being used in a way that clobbers each other (eg `@neutrinojs/style-loader` or `@neutrinojs/web` being used after `@neutrinojs/react`), since it's not migration-specific.

If anyone can think of any other improvements that can be made either to the migration docs or to the runtime checks, please do say! :-)

**Comment by edmorley** [comment]
A new beta has just been published, `v9.0.0-beta.1`. The changes since beta.0 are:
https://github.com/neutrinojs/neutrino/compare/d48eb35...1c082ac

Please try updating to it and let us know if any issues :-)

**Comment by armenzg** [comment]
I've upgraded:
https://github.com/mozilla-frontend-infra/firefox-health-dashboard/pull/194

After upgrading when I run the tests I got the following message [1].
Thanks for the checks! I would have never noticed and remove the links section.

[1]
```
yarn run v1.10.1
$ jest

An error occurred when loading the Neutrino configuration.

ConfigurationError: The default Neutrino HTML template no longer supports the "links" option. To set a favicon use the "favicon" option instead. For stylesheets either import the equivalent npm package from JS, or if using a CDN is preferred, then copy the default HTML template into your repository (where it can be customised as desired) and set the "template" option to its path.
    at module.exports (/Users/armenzg/repos/health/frontend/node_modules/@neutrinojs/html-template/index.js:6:11)
    at Neutrino.use (/Users/armenzg/repos/health/frontend/node_modules/neutrino/Neutrino.js:211:7)
    at Object.entries.forEach (/Users/armenzg/repos/health/frontend/node_modules/@neutrinojs/web/index.js:162:18)
    at Array.forEach (<anonymous>)
    at module.exports (/Users/armenzg/repos/health/frontend/node_modules/@neutrinojs/web/index.js:157:6)
    at Neutrino.use (/Users/armenzg/repos/health/frontend/node_modules/neutrino/Neutrino.js:211:7)
    at module.exports (/Users/armenzg/repos/health/frontend/node_modules/@neutrinojs/react/index.js:43:12)
    at Neutrino.use (/Users/armenzg/repos/health/frontend/node_modules/neutrino/Neutrino.js:211:7)
    at Neutrino.use (/Users/armenzg/repos/health/frontend/node_modules/neutrino/Neutrino.js:216:12)
    at Neutrino.use (/Users/armenzg/repos/health/frontend/node_modules/neutrino/Neutrino.js:220:12)
error Command failed with exit code 1.
```

**Comment by edmorley** [comment]
> There's a bug with `yarn create` and using a version number

https://github.com/yarnpkg/yarn/issues/6587

**Comment by edmorley** [comment]
@eliperelman thank you for the recent reviews :-)

I don't suppose you could publish a new pre-release? Perhaps as `9.0.0-rc.0` (with tag `next`)?
(Since the last of the significant breaking changes have landed)

Changes since `9.0.0-beta.1` are:
https://github.com/neutrinojs/neutrino/compare/1c082ac...master

**Comment by edmorley** [comment]
`v9.0.0-rc.0` has now been released - please could everyone give it a go? :-)

Changelog:
[v9.0.0-beta.1 -> v9.0.0-rc.0](https://github.com/neutrinojs/neutrino/compare/1c082ac...e9c21b8)

**Comment by yordis** [comment]
I am not sure it this preset it helpful to add it to your list

```javascript
const PnpWebpackPlugin = require('pnp-webpack-plugin');

module.exports = (neutrino, opts = {}) => {
  // Adds support for installing with Plug'n'Play, leading to faster installs and adding
  // guards against forgotten dependencies and such.
  neutrino.config.resolve
    .plugin('pnp')
    .use(PnpWebpackPlugin)
    .end()
    .end()
    // Also related to Plug'n'Play, but this time it tells Webpack to load its loaders
    // from the current package.
    .resolveLoader.plugin('pnp')
    .use(PnpWebpackPlugin.moduleLoader(module));
};
```

**Comment by yordis** [comment]
@edmorley before officially release Neutrino 9, since introduce breaking changes. I want to share some thoughts with you.

Since `use` ends up on this https://github.com/neutrinojs/neutrino/blob/a19034e454698a933bc8fb91a2eb44e454691e58/packages/neutrino/Neutrino.js#L211

Do you think that this is an indirection https://github.com/neutrinojs/neutrino/blob/a19034e454698a933bc8fb91a2eb44e454691e58/packages/neutrino/Neutrino.js#L212-L241 from the API perspective?

One detail about middlewares/presets, they take the `options` as a second parameter, `options` that most likely you pass it when you add the middleware to your configuration but since you add extra complexity you need to make the middleware/preset (I will use middleware from now on).

Maybe if changing the shape of the middleware to

```typescript
type Middleware = (neutrino: Neutrino)=> void
```

For example

```javascript
const styleMiddleware = (neutrino)=> {
  // do something with the instance
}
```

This would make the middleware easy to understand and when it comes to the usage of the middleware is easier to reason what is going on.

```javascript
neutrino.use(styleMiddleware);
```

From this example, the `options` went await and that is because you can always create a closure for your config.

```javascript
const styleMiddleware = (opts = {})=> (neutrino)=> {
  // do something with the opts
}
```

So the initialization is

```javascript
neutrino.use(styleMiddleware({
  // options
})
```

Since `.neutrinorc.js` is a JavaScript file things like sharing configs between middlewares https://github.com/neutrinojs/neutrino/blob/a19034e454698a933bc8fb91a2eb44e454691e58/packages/neutrino/Neutrino.js#L231 it is a matter of simply create variables that hold the config.

I am not sure if I even want people to pass strings because you load the middleware over here https://github.com/neutrinojs/neutrino/blob/a19034e454698a933bc8fb91a2eb44e454691e58/packages/neutrino/Neutrino.js#L216

People can't find where the dependencies are being used and we can't even get type check without doing hacking things like here https://github.com/straw-hat-team/neutrino/blob/6da2e3c62947a25d94b2ce6e6ed4e0a5f436c401/src/index.ts#L61 where I have to cast my configs because of the extra indirection added in the function. DevTools that keep track of your dependencies and rely on static analysis can't figure out who is using your deps without being aware of how Neutrino works.

What do you think about it?

This is all based on my personal experiences and to be honest I see more downsides (especially that I had battles with people explaining how simple Neutrino itself is) by adding complixity on something that is simply so I want to hear your thoughts about it before you publish Neutrino 9 since you will introduce breaking changes anyway; and in term of fixing the current presets, I am more than happy to support you with the required changes.

**Comment by edmorley** [comment]
Re the last comment - PR #1246 is now open for that and discussions happening over there :-)

Re the comment before about Yarn plug and play - yeah I'd like to add support, though I've had several issues that blocked usage on Windows in particular:
yarnpkg/yarn#6493
yarnpkg/yarn#6683

In addition, we'd need to request that several of Neutrino's packages get added to the whitelist (like react-scripts was for CRA) due to limitations in ESLint, otherwise linting won't work regardless. It's unknown whether yarn will accept such changes.

**Comment by edmorley** [comment]
> `v9.0.0-rc.0` has now been released - please could everyone give it a go? :-)

Last call for people to try it out/file bugs!

**Comment by edmorley** [comment]
We really need to get this release out as soon as possible (ideally in the next few days since I'm away after). There are issues that people are hitting with Neutrino 8 (eg #1259, #1249 and #1066) that have been fixed for some time in Neutrino 9. It would be a shame to lose a bunch of potential users due to a poor first-run experience from using the older version.

@eliperelman do you know when you might be able to update #1246 to be non-breaking? We can always punt on the migrate tool improvements initially, to save time (we can add it before removing support for the old form in eg Neutrino 10).

**Comment by Oblium** [comment]
Using npx has given me an error 

![image](https://user-images.githubusercontent.com/3975657/50222270-12fd1900-0398-11e9-88bc-fb9dbc434fe4.png)

Instead, using yarn it has worked smoothly.
For system information refer to issue #1259

**Comment by amorriscode** [comment]
I was trying to use the `Components` option of v9 and noticed that it failed when doing a one time lint.

```
⏳  Performing one-time lint
✖ Error: The command "yarn lint --fix" exited unsuccessfully. Try again with the --debug flag for more detailed information about the failure.
    at Project._spawnSync (/Users/amorris2/.npm/_npx/27915/lib/node_modules/@neutrinojs/create-project/commands/init/index.js:41:9)
    at Project.install (/Users/amorris2/.npm/_npx/27915/lib/node_modules/@neutrinojs/create-project/commands/init/index.js:247:12)
    at Object.<anonymous> (/Users/amorris2/.npm/_npx/27915/lib/node_modules/@neutrinojs/create-project/node_modules/yeoman-generator/lib/index.js:424:27)
    at /Users/amorris2/.npm/_npx/27915/lib/node_modules/@neutrinojs/create-project/node_modules/run-async/index.js:25:25
    at new Promise (<anonymous>)
    at /Users/amorris2/.npm/_npx/27915/lib/node_modules/@neutrinojs/create-project/node_modules/run-async/index.js:24:19
    at self.env.runLoop.add.completed (/Users/amorris2/.npm/_npx/27915/lib/node_modules/@neutrinojs/create-project/node_modules/yeoman-generator/lib/index.js:425:13)
    at runCallback (timers.js:810:20)
    at tryOnImmediate (timers.js:768:5)
    at processImmediate [as _immediateCallback] (timers.js:745:5)
```

It worked fine when selecting `A web or Node.js application`.

**Comment by edmorley** [comment]
> Using npx has given me an error

@Oblium, hi! Thank you for reporting that error. It seems that it might be related to a broken npm/npx install - eg:
https://github.com/jxnblk/ram/issues/8

...could you try uninstalling/reinstalling latest npm?

**Comment by edmorley** [comment]
@amorriscode, I believe this might be #1261 which is fixed on master but not yet in a new Neutrino 9 pre-release. Until one is released a workaround is to use Yarn instead.

**Comment by amorriscode** [comment]
@edmorley I'm sure I'm doing something wrong but I have the same issue with yarn. This is how I went about doing it:
`~ yarn global add @neutrinojs/create-project@next`
`~ create-project test-yarn-next`

Here is the error:
```
⏳  Performing one-time lint
✖ Error: The command "yarn lint --fix" exited unsuccessfully. Try again with the --debug flag for more detailed information about the failure.
    at Project._spawnSync (/Users/amorris2/.config/yarn/global/node_modules/@neutrinojs/create-project/commands/init/index.js:41:9)
    at Project.install (/Users/amorris2/.config/yarn/global/node_modules/@neutrinojs/create-project/commands/init/index.js:247:12)
    at Object.<anonymous> (/Users/amorris2/.config/yarn/global/node_modules/yeoman-generator/lib/index.js:424:27)
    at /Users/amorris2/.config/yarn/global/node_modules/run-async/index.js:25:25
    at new Promise (<anonymous>)
    at /Users/amorris2/.config/yarn/global/node_modules/run-async/index.js:24:19
    at self.env.runLoop.add.completed (/Users/amorris2/.config/yarn/global/node_modules/yeoman-generator/lib/index.js:425:13)
    at runCallback (timers.js:810:20)
    at tryOnImmediate (timers.js:768:5)
    at processImmediate [as _immediateCallback] (timers.js:745:5)
```

**Comment by Oblium** [comment]
> > Using npx has given me an error
> 
> @Oblium, hi! Thank you for reporting that error. It seems that it might be related to a broken npm/npx install - eg:
> [jxnblk/ram#8](https://github.com/jxnblk/ram/issues/8)
> 
> ...could you try uninstalling/reinstalling latest npm?

Hi! Updating node, npm and npx it works as expected 😎 
Great work!

**Comment by edmorley** [comment]
@amorriscode oh sorry about. We have a few more fixes on master that will improve the log output in this case, which we hope to get released soon. I'll let you know when that happens.

**Comment by bingomanatee** [comment]
I am unclear how the react/web html template system is now. How do I add links etc., into my index.html file? I've seen references that html-template is not used anymore so... what is?

**Comment by timkelty** [comment]
@bingomanatee it uses https://github.com/jantimon/html-webpack-plugin: https://master.neutrinojs.org/packages/html-template/

**Comment by timkelty** [comment]
@bingomanatee `@neutrinojs/env` was removed in v9 and implemented directly in web presets, so try passing an `env` option to `@neutrinojs/react` with your array.

@edmorley Migration guide should probably point out that `@neutrinojs/env` was removed, yeah?

**Comment by eliperelman** [comment]
> Migration guide should probably point out that `@neutrinojs/env` was removed, yeah?

It appears that this is already in the migration guide.

**Comment by stevehawley** [comment]
I'm using 9.0.0-rc.0, have a small app mostly working (build, start, lint), but testing with jest is only partly working, the simple sanity test runs and passes but jsx tests are having a problem.

I'm trying to write a jest enzyme shallow render test in a .jsx file and it is complaining that it encountered an unexpected token when it gets to a jsx expression.

I didn't see in the migration guide any additional configuration needed to get babel working on tests, and the docs said it should pick up the project wide babel config but the sources are compiling and the tests aren't.

Is this expected to work out of the box on a react project? Did I miss some configuration step in the docs?

Any help appreciated!

**Comment by senpost** [comment]
Any idea of release date for V9? Thank you.

**Comment by Mihai-Ionut-Aurel** [comment]
Hi. I am having some issues upgrading from V8 to V9. I am not that experienced with front-end technologies so it might be that this issue is quite simple to fix.  I can't seem to get Bebpack to resolve the node_modules folder in a proper way. Looking at the v8 and v9 generated config, the folders are missing. Am I missing a step in the migration guide?
I am using an already made [template with React and Play Framework](https://github.com/setusoft/silhouette-play-react-seed which I am trying to update to the latest version. The configuration that the app uses can be found in the [app-ui folder.](https://github.com/setusoft/silhouette-play-react-seed/tree/master/app-ui).
Thank you for the help!

**Comment by davidje13** [comment]
I'm experimenting with migrating a fairly new codebase over to neutrino 9.

So far I haven't found any actual bugs, but the migration guide leaves a **lot** to be desired. It is not a pleasant experience. Here are some of the issues I have encountered so far:

* I have to add dependencies for `eslint`, `jest`, `webpack`, `webpack-cli` and `webpack-dev-server` to my package.json. I'm sure there would be others if I was using other neutrino plugins, and I must be missing some because the tests still don't run. This change isn't listed in the migration guide. Also I have to use jest 23, not 24 (which admittedly only just came out, but again, not documented)

  * sidenote: why are these now peer dependencies rather than pulled in automatically?

* The suggested scripts (start/test/lint/build) are a big help, but they seem to be only for frontend applications. I have a backend application (using `@neutrinojs/node`) and I have *no idea* what to write as a `start` script to get equivalent behaviour to what I had before. Also I think the start script shouldn't contain `--open`, because the old behaviour did not do that.

* The automatic `@neutrinojs/copy` step being removed is fine, but the update guide just points to the PR for it, which doesn't show how to migrate! I've added [a comment to that issue](https://github.com/neutrinojs/neutrino/pull/814#issuecomment-466821552) which contains what I think is the correct configuration to maintain behaviour, but I had to look at the changed files to figure that out! Remember that users aren't all familiar with every neutrino package and how to configure it.

* In `@neutrinojs/react`, support for `html.headHtmlSnippet` has disappeared (maybe? maybe renamed?). With `html.links` also vanishing, this leaves me with nowhere to put my `apple-touch-icon` or `manifest` links. What do I do now? And where do I put my meta overrides such as `<meta name="apple-mobile-web-app-capable" content="yes">`?

* I can't get Jest working. I'm sure it's something I'm missing, but it won't run through webpack before running the tests, so it chokes on ES6 and JSX. It says:

    > By default, if Jest sees a Babel config, it will use that to transform your files, ignoring "node_modules".

    So I guess this means that the react and node packages used to auto-configure some babel stuff? *what was it?*

All these can be fixed by delving into source code / running debugging with the old version (although I'm not sure where to begin on the jest one). But the point of a migration guide is that I *shouldn't have to do that*.

As I say, no bugs that I've seen so far (unless something I've listed is actually a bug), but I don't think the migration guide is ready for prime time. There will be a lot of very confused and frustrated users.

**Comment by davidje13** [comment]
A non-documented breaking change for `@neutrinojs/react`:

```javascript
module.exports = {
  options: {
    port: 1234,
  },
// [...]
}
```

Does not work any more. Instead it must be:

```javascript
// [...]
    ['@neutrinojs/react', {
      devServer: {
        port: 1234,
      },
    }],
// [...]
```

**Comment by edmorley** [comment]
Sorry for the lack of replies recently - I'm mid changing jobs so pretty busy at the moment, however should have some more time from the middle of next week onwards :-)

**Comment by flyfishMT** [comment]
Hi, 

I am new to neutrino but just installed the next version and am trying to create a component lib and am receiving an error. I ran with the debug flag but it didn't look like it added any useful info.

Also, I'm not sure why it's using yarn as I use npm.

Thought you might want to know.  Thanks

```
c:/pwd>npx @neutrinojs/create-project@next packages/common-ui
npx: installed 347 in 13.308s
                          _          _
      _ __    ___  _   _ | |_  _ __ (_) _ __    ___
     | '_ \  / _ \| | | || __|| '__|| || '_ \  / _ \
     | | | ||  __/| |_| || |_ | |   | || | | || (_) |
     |_| |_| \___| \__,_| \__||_|   |_||_| |_| \___/

Welcome to Neutrino! ??
To help you create your new project, I am going to ask you a few questions.

? ??  First up, what would you like to create? Components
? ??  Next, what kind of components would you like to create? React Components
? ??  Would you like to add a test runner to your project? Jest
? ??  Would you like to add linting to your project? Airbnb style rules

??  Looks like I have all the info I need. Give me a moment while I create your project!

   create common-ui\package.json
   create packages\common-ui\.neutrinorc.js
   create packages\common-ui\webpack.config.js
   create packages\common-ui\src\index.jsx
   create packages\common-ui\src\components\HelloWorld\index.jsx
   create packages\common-ui\jest.config.js
   create packages\common-ui\test\simple_test.js
   create packages\common-ui\.eslintrc.js

?  Installing devDependencies: @neutrinojs/airbnb@^9.0.0-rc.0, @neutrinojs/jest@^9.0.0-rc.0, @neutrinojs/react-components@^9.0.0-rc.0, eslint@^5, jest@^23, neutrino@^9.0.0-rc.0, prop-types@^15, react-dom@^16, react@^16, webpack-cli@^3, webpack-dev-server@^3, webpack@^4
?  Performing one-time lint
× Error: The command "yarn lint --fix" exited unsuccessfully. Try again with the --debug flag for more detailed information about the failure.
    at Project._spawnSync (c:\userdir\AppData\Roaming\npm-cache\_npx\2272\node_modules\@neutrinojs\create-project\commands\init\index.js:41:9)
    at Project.install (c:\userdir\AppData\Roaming\npm-cache\_npx\2272\node_modules\@neutrinojs\create-project\commands\init\index.js:247:12)
    at Object.<anonymous> (c:\userdir\AppData\Roaming\npm-cache\_npx\2272\node_modules\@neutrinojs\create-project\node_modules\yeoman-generator\lib\index.js:424:27)
    at c:\userdir\AppData\Roaming\npm-cache\_npx\2272\node_modules\@neutrinojs\create-project\node_modules\run-async\index.js:25:25
    at new Promise (<anonymous>)
    at c:\userdir\AppData\Roaming\npm-cache\_npx\2272\node_modules\@neutrinojs\create-project\node_modules\run-async\index.js:24:19
    at self.env.runLoop.add.completed (c:\userdir\AppData\Roaming\npm-cache\_npx\2272\node_modules\@neutrinojs\create-project\node_modules\yeoman-generator\lib\index.js:425:13)
    at runCallback (timers.js:705:18)
    at tryOnImmediate (timers.js:676:5)
    at processImmediate (timers.js:658:5)
```

**Comment by eliperelman** [comment]
@flyfishMT would you mind sharing some of your environment details such as Node.js and npm versions?

**Comment by flyfishMT** [comment]
Sorry, added. Let me know if you need anything else.

**Comment by Tassfighter** [comment]
> Hi,
> 
> I am new to neutrino but just installed the next version and am trying to create a component lib and am receiving an error. I ran with the debug flag but it didn't look like it added any useful info.
> 
> Also, I'm not sure why it's using yarn as I use npm.
> 
> Thought you might want to know. Thanks
> 
> Environment:
> 
> ```
> Windows 10 | 1803
> npm | 6.9.0
> node | v10.15.1
> ```
> ```
> c:/pwd>npx @neutrinojs/create-project@next packages/common-ui
> npx: installed 347 in 13.308s
>                           _          _
>       _ __    ___  _   _ | |_  _ __ (_) _ __    ___
>      | '_ \  / _ \| | | || __|| '__|| || '_ \  / _ \
>      | | | ||  __/| |_| || |_ | |   | || | | || (_) |
>      |_| |_| \___| \__,_| \__||_|   |_||_| |_| \___/
> 
> Welcome to Neutrino! ??
> To help you create your new project, I am going to ask you a few questions.
> 
> ? ??  First up, what would you like to create? Components
> ? ??  Next, what kind of components would you like to create? React Components
> ? ??  Would you like to add a test runner to your project? Jest
> ? ??  Would you like to add linting to your project? Airbnb style rules
> 
> ??  Looks like I have all the info I need. Give me a moment while I create your project!
> 
>    create common-ui\package.json
>    create packages\common-ui\.neutrinorc.js
>    create packages\common-ui\webpack.config.js
>    create packages\common-ui\src\index.jsx
>    create packages\common-ui\src\components\HelloWorld\index.jsx
>    create packages\common-ui\jest.config.js
>    create packages\common-ui\test\simple_test.js
>    create packages\common-ui\.eslintrc.js
> 
> ?  Installing devDependencies: @neutrinojs/airbnb@^9.0.0-rc.0, @neutrinojs/jest@^9.0.0-rc.0, @neutrinojs/react-components@^9.0.0-rc.0, eslint@^5, jest@^23, neutrino@^9.0.0-rc.0, prop-types@^15, react-dom@^16, react@^16, webpack-cli@^3, webpack-dev-server@^3, webpack@^4
> ?  Performing one-time lint
> × Error: The command "yarn lint --fix" exited unsuccessfully. Try again with the --debug flag for more detailed information about the failure.
>     at Project._spawnSync (c:\userdir\AppData\Roaming\npm-cache\_npx\2272\node_modules\@neutrinojs\create-project\commands\init\index.js:41:9)
>     at Project.install (c:\userdir\AppData\Roaming\npm-cache\_npx\2272\node_modules\@neutrinojs\create-project\commands\init\index.js:247:12)
>     at Object.<anonymous> (c:\userdir\AppData\Roaming\npm-cache\_npx\2272\node_modules\@neutrinojs\create-project\node_modules\yeoman-generator\lib\index.js:424:27)
>     at c:\userdir\AppData\Roaming\npm-cache\_npx\2272\node_modules\@neutrinojs\create-project\node_modules\run-async\index.js:25:25
>     at new Promise (<anonymous>)
>     at c:\userdir\AppData\Roaming\npm-cache\_npx\2272\node_modules\@neutrinojs\create-project\node_modules\run-async\index.js:24:19
>     at self.env.runLoop.add.completed (c:\userdir\AppData\Roaming\npm-cache\_npx\2272\node_modules\@neutrinojs\create-project\node_modules\yeoman-generator\lib\index.js:425:13)
>     at runCallback (timers.js:705:18)
>     at tryOnImmediate (timers.js:676:5)
>     at processImmediate (timers.js:658:5)
> ```

I am getting the same error on:
Linux Ubuntu 16.04
npm 6.7.0
node v8.9.1

**Comment by edmorley** [comment]
> So far I haven't found any actual bugs, but the migration guide leaves a **lot** to be desired. It is not a pleasant experience. Here are some of the issues I have encountered so far:

Hi @davidje13! I agree there are a lot of breaking changes and a lot of having to read between the lines of the migration guide - sorry it's been so frustrating!

>     * I have to add dependencies for `eslint`, `jest`, `webpack`, `webpack-cli` and `webpack-dev-server` to my package.json. I'm sure there would be others if I was using other neutrino plugins. This change isn't listed in the migration guide.

I think the intention was for this to be covered by the "Refer to the updated package installation instructions on your preset's documentation page." mention. I'd like to avoid copy-pasting every preset's setup into the migration guide -- but perhaps there are other things we can do to make this more obvious? (eg making that sentence bold?).

>     * Also I have to use jest 23, not 24 (which admittedly only just came out, but again, not documented)

This is resolved on `master` - I'd like to do an RC1 release soon.

>     * sidenote: why are these now peer dependencies rather than pulled in automatically?

In Neutrino 9, Neutrino is now a configuration generator, whose output is consumed by the CLI tools directly, rather than Neutrino invoking them via their API (see #852 for the issues this resolved). In order for users to interact with the tools directly, they need to be top-level dependencies in their `package.json`, and so peer dependencies of the Neutrino presets.

>     * The suggested scripts (start/test/lint/build) are a big help, but they seem to be only for frontend applications. I have a backend application (using `@neutrinojs/node`)

Yeah it would be good to document this case. #1188 is open for that.

>     * The automatic `@neutrinojs/copy` step being removed is fine, but the update guide just points to the PR for it, which doesn't show how to migrate

I'll add an example to the migration guide.

>     * In `@neutrinojs/react`, support for `html.headHtmlSnippet` has disappeared (maybe? maybe renamed?).

Yeah the template changed, documented here:
https://github.com/neutrinojs/neutrino/blame/3dbc0e44d815c7762bb68483b581c760f67bdbf9/docs/migration-guide.md#L304-L307
(There are quite possibly ways to improve the wording; ideas welcome :-))

>     * After looking through [somebody else's upgrade](https://github.com/mozilla-frontend-infra/firefox-health-dashboard/pull/158/files) I found that I need to create `jest.config.js`:
>       ...
>       Only webpack and eslint are called out in the migration guide; it would be good to at least list the config filenames for all tools neutrino uses (e.g. "webpack -> webpack.config.js, eslint -> .eslintrc.js, jest -> jest.config.js, [...others]"), even if we're left to figure out the content ourselves.

The idea was that this was covered by the "Refer to the updated package installation instructions on your preset's documentation page." reference mentioned above - with the thinking that it would be too verbose to duplicate every preset's installation instructions in the migration guide.

> A non-documented breaking change for `@neutrinojs/react`:
> Also a similar one for disabling hot reload

I'll add these to the migration guide and see about adding more helpful error messages in code too.

---

Overall, I think the best way to simplify the migration process is a combination of:
* adding additional in-code error messages to catch old options/... and provide suitable advice
* expanding the new `migrate` package to handling more changes than just the ones in #1246.
* tweaking the migration guide wording/order and giving more examples

**Comment by davidje13** [comment]
@edmorley thanks for the reply!

I think the "Refer to the updated package installation instructions on your preset's documentation page" puts rather a lot of work in the hands of the updater. Parsing changes from installation instructions involves checking through *everything in the installation guide*, comparing it to what you already have to figure out the delta, and back-filling the explanation / understanding. It also involves figuring out which modules are being used (e.g. `react` brings in `web` brings in `html template`), which is often not obvious and the only reliable way to know is to use `npm ls` or similar.

I would strongly advocate for copying the relevant changes for all packages into the migration guide. If that makes the migration guide huge, then really that just indicates that it is a huge migration (which perhaps means a staged release would help?). Maybe to reduce the massive-document-shock you could split it into separate migration guides for each module a person is using, but again due to it not being obvious which modules are being used transitively, I would advocate for a single (i.e. searchable) page with everything (of course, both versions could be offered, which is often the case with API documentation).

For me, a migration guide says "if you were doing X you will now need to do Y. To learn more, check link Z". Right now, I think this is more "if you were doing X then start exploring the documentation in Y and figure out what you need to do now". That's fine for one or two changes, and certainly better than nothing, but it's incredibly frustrating for dozens of changes and makes the migration into a multi-day task.

**Comment by edmorley** [comment]
> It also involves figuring out which modules are being used (e.g. `react` brings in `web` brings in `html template`), which is often not obvious and the only reliable way to know is to use `npm ls` or similar.

Ah so the installation guide only needs to be followed for the top-level preset (ie `@neutrinojs/react`, `@neutrinojs/airbnb`, `@neutrinojs/jest`) - which are the packages that users have listed in their `package.json` rather than the lower level presets too.

I agree with the rest of what you're saying though.

The situation we're in is that Neutrino 9 has a lot of breaking changes, and the migration process is likely frustrating for people who aren't core developers and/or who haven't used earlier betas, but the longer we leave it before releasing, the more people who will have to go through this process (and v8 has quite a few bugs). As such I think we'll have to do a few quick fixes, release v9 and see about working on improving the migrate tool after (#1336).

**Comment by edmorley** [comment]
> Parsing changes from installation instructions involves checking through _everything in the installation guide_, comparing it to what you already have to figure out the delta, and back-filling the explanation / understanding.

For v9, pretty much all of the installation process has changed, so I think the delta is pretty much "everything in the installation section" rather than just some of it? (It it was only a small percentage, I think cherry-picking the changes into the migration guide would definitely make sense.)

I have a WIP for some changes to the migration guide that might help a bit for now - I hope to have it open later today.

**Comment by edmorley** [comment]
> Also a similar one for disabling hot reload (this is omitted in the migration guide and shows old incorrect information in the react readme): setting `hot` to false won't work any more. A way to disable hot reloading now is:

So this one is tricky, since technically it's working as expected - setting `hot: false` stops the addition of react-hot-loader, and disables webpack's hot module reloading mode. However HMR being disabled doesn't mean the page won't reload - just that it does a full reload, rather than a "hot" reload.

**Comment by davidje13** [comment]
> So this one is tricky, since technically it's working as expected - setting `hot: false` stops the addition of react-hot-loader, and disables webpack's hot module reloading mode. However HMR being disabled doesn't mean the page won't reload - just that it does a full reload, rather than a "hot" reload.

I guess the interesting part is that this was not the v8 behaviour then; setting `hot: false` prevented any reload (hot or full refresh). Worth documenting both versions, since I can see people in different situations favouring both. Maybe a future feature could make it an enum; `reload: 'hot'` / `reload: 'refresh'` / `reload: 'none'`.

**Comment by edmorley** [comment]
Hi! I've just released `v9.0.0-rc.1`.

Of note is that the syntax for using presets/middleware in `.neutrinorc.js` has changed, which will require either running the new migrate tool or manual changes to adjust. See:
https://master.neutrinojs.org/migration-guide/#neutrino-v8-to-v9
https://master.neutrinojs.org/migrate/

The full changelog can be found here:
https://github.com/neutrinojs/neutrino/compare/e9c21b8...v9.0.0-rc.1

**Comment by timkelty** [comment]
Test driving now!

**Comment by amorriscode** [comment]
@edmorley just curious when you think another release candidate might be released since @timkelty fixed a bug that is currently blocking (#1369).

**Comment by eliperelman** [comment]
@amorriscode I am waiting for the latest lockfile changes before releasing, should be soonish.

**Comment by amorriscode** [comment]
Hey @eliperelman, sorry to be a pain but thought I'd check in. Still blocked by #1369. 😀

**Comment by davidje13** [comment]
@amorriscode the short-term workaround for that issue is to specify an exact version for core-js rather than a version range. For example I have set `3.0.1` in my code. You can switch back to a range once the patch is released. That should prevent this from being a blocking issue.

**Comment by eliperelman** [comment]
I just released `v9.0.0-rc.2`.

[View Changelog](https://github.com/neutrinojs/neutrino/compare/v9.0.0-rc.1...9.0.0-rc.2)

**Comment by eliperelman** [comment]
Aaaaaaand I just released `v9.0.0-rc.2` due to a bug in create-project.

[View Changelog](https://github.com/neutrinojs/neutrino/compare/9.0.0-rc.2...9.0.0-rc.3)

**Comment by syncharny** [comment]
Any idea why I'm running across this error?


```
node_modules/neutrino/bin/neutrino.js --inspect --mode production

An error occurred when loading the Neutrino configuration.

ConfigurationError: @neutrinojs/web is being used when a `compile` rule already exists, which would overwrite the existing configuration. If you are including this preset manually to customise rules configured by another preset, instead use that preset's own options to do so.
    at neutrino (app-ui/node_modules/@neutrinojs/web/index.js:15:11)
    at Neutrino.use (app-ui/node_modules/neutrino/Neutrino.js:182:26)
    at neutrino (app-ui/node_modules/@neutrinojs/react/index.js:56:12)
    at Neutrino.use (app-ui/node_modules/neutrino/Neutrino.js:182:26)
    at use.forEach.use (app-ui/node_modules/neutrino/index.js:49:37)
    at Array.forEach (<anonymous>)
    at module.exports (app-ui/node_modules/neutrino/index.js:49:13)
    at Object.<anonymous> (app-ui/node_modules/neutrino/bin/neutrino.js:9:3)
    at Module._compile (internal/modules/cjs/loader.js:701:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:712:10)
```


neutrino versions:
```
"@neutrinojs/copy": "^9.0.0-rc.3",
    "@neutrinojs/karma": "^9.0.0-rc.3",
    "@neutrinojs/react": "^9.0.0-rc.3",
    "neutrino": "^9.0.0-rc.3",
```

and my .neutrinorc:

```
const path = require('path');
const HtmlIncludeAssetsPlugin = require("html-webpack-include-assets-plugin");
const webpack = require("webpack");
const react = require('@neutrinojs/react');
const htmlTemplate = require('@neutrinojs/html-template')

module.exports = {
  use: [
    react(),
    react(
      {
        devServer: {
          headers: {
            "Access-Control-Allow-Origin": "*",
          }
        },
        babel: {
          presets: [
            '@lingui/babel-preset-react',
          ],
        },
      }
    ),
    htmlTemplate(
      {
        title: 'Test App',
        mobile: true,
        favicon: [
          'https://fonts.googleapis.com/css?family=Roboto:400,500,700',
          {
            href: '/static/favicon.ico',
            rel: 'shortcut icon',
            type: 'image/x-icon'
          },
        ],
      }
    )
  ],
  options: {
    output: path.resolve(__dirname, '../target/npm/dist/ui')
  }
};
```

**Comment by syncharny** [comment]
my bad, typo with the duplicate call to react in the rc file.

**Comment by IceMaD** [comment]
Hello, good job wor the work on the v9. 

I had just one little problem while migrating from v8 to v9.
On my project, I use legacy decorators. To make them work, babel plugins order is [important](https://babeljs.io/docs/en/babel-plugin-proposal-decorators) but it seems that neutrino v9 sort the babel plugins by alphabetical order which break the compilation.

I have set up a repo with the minimal code to reproduce the error : https://github.com/IceMaD/neutrino-sort-bug

On master branch, the build fail. On fix branch, the build pass.

Would it be hard to keep the order ?

**Comment by IceMaD** [comment]
Hello, good job wor the work on the v9. 

I had just one little problem while migrating from v8 to v9.
On my project, I use legacy decorators. To make them work, babel plugins order is [important](https://babeljs.io/docs/en/babel-plugin-proposal-decorators) but it seems that neutrino v9 sort the babel plugins by alphabetical order which break the compilation.

I have set up a repo with the minimal code to reproduce the error : https://github.com/IceMaD/neutrino-sort-bug

On master branch, the build fail. On fix branch, the build pass.

Would it be hard to keep the order ?

**Comment by crimx** [comment]
I am writing a preset and want to enhance the eslint preset if exist. Can it be done in Neutrino 9? It seems Neutrino 9 does not offer chainable for other configs.

**Comment by timkelty** [comment]
@crimx you're just trying to pass config to eslint? If so, just pass them like so, anythey will be merged:

```
neutrino.use(
  eslint({
    eslint: {
      baseConfig: {
        // Your config here…
      }
    }
  }),
);
```

**Comment by crimx** [comment]
@timkelty You can't pass options to other presets without wrapping them into your own preset right?

**Comment by crimx** [comment]
A workaround is to hijack the output handler:

```javascript
if (neutrino.outputHandlers.has('eslintrc')) {
    const eslintrc = neutrino.outputHandlers.get('eslintrc')
    neutrino.outputHandlers.set('eslintrc', (...args) => {
      const config = eslintrc(...args)
      // do some changes
      return config
    })
  }
```

**Comment by timkelty** [comment]
@crimx you should always be able to `.tap` options after the fact…
Want to post your whole neutrinorc and what you're after?

**Comment by crimx** [comment]
@timkelty I am writing a preset, there is no neutrinorc.

It's a preset for WebExtension. I need to add a global to eslint config if it exist. I need a way to do this without adding the eslint (or standardjs or airbnb) preset to my dependencies.

**Comment by timkelty** [comment]
@IceMaD try something like this:

```js
const react = require('@neutrinojs/react');

module.exports = {
  options: {
    root: __dirname
  },
  use: [
    react({
      html: {
        title: 'test'
      }
    }),

    neutrino =>
      neutrino.config.module
        .rule('compile')
        .use('babel')
        .tap(options => {
          const plugins = [
            ['@babel/plugin-proposal-decorators', { legacy: true }]
          ];

          options.plugins.unshift(...plugins);

          return options;
        })
  ]
};

```

**Comment by timkelty** [comment]
@crimx I see…something like this should work then (lint presets must be used first):
```
const eslint = require('@neutrinojs/eslint');
const merge = require('deepmerge');

module.exports =  (opts = {}) => neutrino => {
  const lintRule = neutrino.config.module.rules.get('lint');

  if (lintRule) {
    lintRule.use('eslint').tap(lintOptions =>
      merge(lintOptions, {
        baseConfig: {
          globals: {
            MY_GLOBAL: true
          }
        }
      })
    );
  }

  // …the rest of your preset
};

module.exports = {
  use: [eslint(), webExtensionPreset()]
};
```

**Comment by crimx** [comment]
@timkelty This only `tap` the eslint Webpack config, but not the `neutrino().eslintrc()` output which can only be configured via `eslint(options)`. 

This is what I did. It works with a bit of hack.

```javascript
  if (neutrino.outputHandlers.has('eslintrc')) {
    const eslintrc = neutrino.outputHandlers.get('eslintrc')
    neutrino.outputHandlers.set('eslintrc', (...args) => {
      const config = eslintrc(...args)
      if (!config.globals) {
        config.globals = {}
      }
      config.globals.browser = true
      return config
    })
  }

  if (neutrino.config.module.rules.has('lint')) {
    neutrino.config.module
      .rule('lint')
      .use('eslint')
      .tap(options => {
        if (!options.globals) {
          options.globals = []
        }
        options.globals.push('browser')
        return options
      })
  }
```

**Comment by timkelty** [comment]
@crimx hmmm what does you .eslintrc.js look like?

with this:

```
// .eslintrc.js
const neutrino = require('neutrino');

module.exports = neutrino().eslintrc();
```

and running this: 
```
eslint --print-config src/index.js
```

I get the merged config. You're seeing different?

**Comment by crimx** [comment]
@timkelty  Holy moley! Didn't even think it can work this way! I somehow missed these few lines when reading the source:

https://github.com/neutrinojs/neutrino/blob/09f165668aa607731cbdb5ec9bed699ae8441c1c/packages/eslint/index.js#L11-L14

Sorry for wasting your time!

**Comment by timkelty** [comment]
@crimx 😎no prob!

**Comment by armenzg** [comment]
What's left before we release?

**Comment by eddyw** [comment]
Hello,
Not sure if I should raise a new issue or just ask here 🙈 
Will neutrino include official TS typings? When using v8 and installing `webpack-chain`, we got nice typings for at least `neutrino.config.*` (because of `webpack-chain` typings) but with v9 is all just `any` and it's difficult to work with.

I understand that presets/middleware are just JS files but typings still work there, e.g (with v8):
![image](https://user-images.githubusercontent.com/1407526/63943093-8f0c3c80-caa1-11e9-898f-a88f7a55e7cd.png)

Kind of ugly JSDoc but it helped reduce the time I had to spend going over the docs because I forgot the name of a method, for example.

**Comment by zeropaper** [comment]
Hi there!

I'm having some issues with the worker-loader (since it's not included in v9).
My config looks like (borrowed from https://github.com/neutrinojs/neutrino/pull/1069/files#diff-2be40b157961444bdf381c35002b0312L88)
```js
    // …
    react({
      babel: {
        // …
      },
    }),
    (neutrino) => {
      neutrino.config.module
        .rule('worker')
          .test(/\.worker\.js$/)
          .use('worker')
            .loader(require.resolve('worker-loader'))
            .end()
          .end()
        .end();
    },
    // …
```

The loader seems to "load" and "compile" the worker but the compilation seems to add the Webpack (here the code at the beginning of the compiled worker):

```js
/******/ (function(modules) { // webpackBootstrap
/******/ 	var parentHotUpdateCallback = window["webpackHotUpdate"];
/******/ 	window["webpackHotUpdate"] = // eslint-disable-next-line no-unused-vars
/******/ 	function webpackHotUpdateCallback(chunkId, moreModules) {
/******/ 		hotAddUpdateChunk(chunkId, moreModules);
/******/ 		if (parentHotUpdateCallback) parentHotUpdateCallback(chunkId, moreModules);
/******/ 	} ;
/******/
/******/ 	// eslint-disable-next-line no-unused-vars
/******/ 	function hotDownloadUpdateChunk(chunkId) {
/******/ 		importScripts(__webpack_require__.p + "" + chunkId + "." + hotCurrentHash + ".hot-update.js");
/******/ 	}
/******/
// …
```

And obviously, the `window` object (accessed at line 2) is not available on a WebWorker… and throws an error.

Let's face it… I'm kind of lost. Should it be an issue I should fill here… or at the `worker-loader` project?
Any additional guidance on how to solve that would be very welcome!
I would made a PR for the upgrade path in the documentation (if I get fixed).

Cheers!

**Comment by edmorley** [comment]
@zeropaper Hi! Thank you for the clear write up. It looks like the `window` usage is related to the hot reloading mode (HMR) - see https://github.com/webpack-contrib/worker-loader/issues/174. There are a few workarounds listed in that GitHub issue, such as disabling HMR mode. With Neutrino disabling HMR can be achieved by setting `hot: false` in the Web/React/... preset options.

**Comment by edmorley** [comment]
@eddyw Hi! Yeah I can see how everything being of type Any must be frustrating. I personally wouldn't mind converting Neutrino to TypeScript, though I think other maintainers might have different opinions. If there was an easy way to improve types out of the box (whilst sticking to JS) that doesn't impose too much of a burden (eg something checked in CI) - like was performed for webpack-chain, a PR for that would likely be accepted. However if it's more than just a few top-level type definitions it may be worth going the whole way and converting to TS entirely (but that's not something I personally will have time to work on).

**Comment by zeropaper** [comment]
@edmorley I noticed the numerous issues on GH a few minutes after posted (sorry).
Disabling HMR is not really an option for me but I will dig into a few solutions and report here.
Quick question thou… If I manage to get it fixed, should I made a documentation PR on that
https://github.com/neutrinojs/neutrino/blob/master/docs/migration-guide.md ?

**Comment by edmorley** [comment]
@zeropaper yeah a migration docs PR would be most welcome :-)

**Comment by edelhart-voicebase** [comment]
The amount of chaos bugs and bad documentation in the neutrino process has driven us to abandon Neutrino and un-reccomend it to others. Please up your game.

**Comment by edmorley** [comment]
@edelhart-voicebase Hi! Could you be more specific - I've had a quick glance at the issues/PRs in this repository and I can't find any bugs reported by you? As is the above comment isn't very constructive (also see https://flaviocopes.com/developer-bucket/). 

If you're using Neutrino 8 there are definitely known bugs - I'd strongly recommend using Neutrino 9 release candidate instead (ideally Neutrino 9 stable would already be out, but there have been a constant stream of small things that were blocking it). 

Note when using the release candidate, you'll want to use the docs on master (perhaps this is why you think the docs are broken?):
https://master.neutrinojs.org

**Comment by bingomanatee** [comment]
The fact that you haven’t released the version of Neutrino that works says it all. Neutrino is slow, and I created a production admin site in react, that past s certain point, ran out of memory. All my bugs have been reported by other people. FWIW I have replaced Neutrino with:

* Parcel.js for web projects
* Rollup for web libraries and components

and have gotten a lot more responsive work done.

**Comment by edmorley** [comment]
There are no serious bugs still filed against Neutrino 9, so either your issues are something new (in which case please file a bug), or you are referring to issues that existed in Neutrino 8, but that are now fixed in the release candidate. If that latter, I agree that we need to get Neutrino 9 out ASAP, but it's also puzzling why you seem against using it over Neutrino 8.

**Comment by Tassfighter** [comment]
I can't confirm Neutrino is slow and full of chaos bugs. I am running a fast-growing platform with many Neutrino 9 projects and never had major issues.

_If that latter, I agree that we need to get Neutrino 9 out ASAP_ **Thumbs up!**

**Comment by bingomanatee** [comment]
I can’t rely on this project as 9 has been pre-release for over half a year. You’re simply not active enough to be trusted.

**Comment by yordis** [comment]
@bingomanatee actually has been almost a year using Neutrino pre-releases so I can't agree with you, it is a matter of `"understanding"` the differences between Neutrino and the presets.

I don't use any preset since I converted `create-react-app` config and I don't find that much value on the presets.

Despite your valid frustrations @bingomanatee,

I condemn your attitude. You could always take a different approach, a more proactive and supporter. It is an OS project, maybe instead of converting everything to Rollup and Parcel, take the same time and contribute back to the project and help the people to maintain it.

**Comment by timkelty** [comment]
> I can’t rely on this project as 9 has been pre-release for over half a year. You’re simply not active enough to be trusted.

I find it preferable to "trust" a project that has a methodical and deliberate release process, opposed to one with frequent breaking releases.

Also worth noting is that v9 introduces some of the most significant conceptual changes to the project, and thus is a very significant release to get right.

**Comment by akhnaz** [comment]
I am currently using neutrino 8 with react preset but would like to use babel 7. The only way that I know is to upgrade the neutrino to v9 rc. Is that true or I have the ability to change the babel version without migrating to v9?

**Comment by yordis** [comment]
@akhnaz are you using the presets?

**Comment by akhnaz** [comment]
@yordis I am using presets react

**Comment by dantman** [comment]
> @edmorley I would propose something: make Neutrino releases decoupled from the presets.

I don't really care about usage of neutrino without presets, IMHO presets are what make neutrino useful. But I agree that Neutrino's release process for core and presets should be decoupled from each other.

Inside the release process presets seem to slow down the release of bugfixes to core that affect everyone. While outside the main repo presets suffer from maintenance issues, in my experience frequently becoming abandoned making it impossible to fix bugs and limitations in them.

I think presets should stay inside the neutrino mono-repo and we should invite adding more presets to the mono-repo. But the release process should be decoupled so that core and individual presets can get releases without waiting for a bug in a separate preset to be fixed.

Then instead of presets becoming abandoned and hard to fix when one person no-longer finds them useful, anyone using them can fix bugs simply by submitting a PR that can be merged by someone other than only the original author. And core can get releases more frequently without waiting for any one of the presets to have all its bugs fixed.

> I find it preferable to "trust" a project that has a methodical and deliberate release process, opposed to one with frequent breaking releases.

Personally I find it's usually the opposite. Most projects I see that make frequent breaking changes (Material UI, React Native) make small near-painless breaking changes that make migration easy and usually because of it have good tooling (codemods and compat shims) to make those migrations easier. While projects that avoid making breaking changes for a long time and wait to do it all at once in a long process (Babel) draw out the migration to a painful length and lead to a hell within the community where people are on different versions, stable releases become unusable bug-ridden problems, alphas become the de-facto "stable" version, and migration is painful.

**Comment by timkelty** [comment]
> Personally I find it's usually the opposite…

@dantman You certainly have a point there, especially with the projects you mentioned. Somewhere in the middle is likely ideal.

**Comment by timkelty** [comment]
Has anyone else played around with multi-compiler output w/ v9?

I'm attempting to use a single `.neutrinorc.js`, but compile 2 outputs: 1 esmodules, 1 without (legacy).

Here's what I have so far…

webpack.config.js
```
const neutrino = require('neutrino');
const {join} = require('path');
const rc = require(join(process.cwd(), '.neutrinorc.js'));
const esmOptions = {...rc.options, ...{
  esm: true,
  mains: {
    'index.esm': rc.options.mains.index,
  },
}};

module.exports = [
  neutrino(rc).webpack(),
  neutrino({...rc, ...{options: esmOptions}}).webpack(),
];
```

…then in my `.neutrinorc.js` file (abbreviated for…brevity):

```
const web = require('@neutrinojs/web');

module.exports = {
  options: {
    mains: {
      index: 'index',
    },
  },
  use: [
    (neutrino) => neutrino.use(
      web({
        // Pass `false` to use .browserslistrc file
        targets: neutrino.options.esm ? {esmodules: true} : false,
      })
    ),
  ],
};
```

_seems_ to be working? Curious if anyone else has done this, or if this seems like a reasonable approach?

**Comment by edmorley** [comment]
@timkelty looks good to me - it's the approach we mention here:
https://master.neutrinojs.org/usage/#generating-multiple-builds

**Comment by timkelty** [comment]
@edmorley  Ah - totally missed that in the guide, thanks!

I guess the difference from my example is I'm creating 2 full neutrino instances with different options, rather than just manipulating the webpack output.

I wasn't sure if setting an arbitrary option on `neutrino.options` would work (I feel like maybe it used to validate those?), but it seems to allow it.

**Comment by edmorley** [comment]
I've just published `v9.0.0-rc.4`. Most notably it includes support for ESLint 6.

Full changes here:

**Comment by edmorley** [comment]
I'd like to ship Neutrino 9 final in the next week or two. If anyone sees issues with RC4 please file them as soon as possible. Also if anyone has spare time to work on #1404 it would be appreciated :-)

**Comment by armenzg** [comment]
I've deployed it to a couple of projects. So far no problems. Thanks Ed!

**Comment by edmorley** [comment]
@armenzg That's great! Hope you are well and things going great for Treeherder and the rest of the team :-)

**Comment by mlnmln** [comment]
Thank you so much for the [migration tool](https://master.neutrinojs.org/migrate/) and your work on the project. 

Currently exploring `9.0.0-rc.4`, so far relatively smooth ride. 

Since `@neutrinojs/stylelint` is going to be deprecated, what's the suggested upgrade path here? 

I would have expected something similar as for the eslint package, e.g. 

```
// .stylelintrc.js
const neutrino = require("neutrino");

// Load with no config overrides:
module.exports = neutrino().stylelintrc();
```

**Comment by Tassfighter** [comment]
Any news on the release date?

**Comment by edmorley** [comment]
I've just released v9.0.0-rc.5:
https://github.com/neutrinojs/neutrino/blob/master/CHANGELOG.md#v900-rc5

Presuming no issues reported in the next few days, I'd like to release v9 final at that point.

**Comment by ringods** [comment]
@edmorley I only see a bump of a few dependencies since 9.0.0-rc5. Together with it, the only issue reported since you released 9.0.0-rc5 is #1492, but that doesn't seem to be related to the release candidate per se.

Time to finally make that 9.0.0 release? 😉

**Comment by edmorley** [comment]
Hi! Yeah #1492 is the only potential blocker (not in the sense that it's a regression, but because it's a breaking change) - though perhaps it's best to just leave it until Neutrino 10 now.

**Comment by jeanlescure** [comment]
Just jumped on the Neutrino band-wagon. I struggled with v8 until I found these threads pointing to v9 RC.

I would also like to echo the overall sentiment, please release 9.0.0 😄

**Comment by edmorley** [comment]
I've opened #1505 to begin the v9.0.0 final release process :-)

**Comment by edmorley** [comment]
Neutrino 9.0.0 stable is now released. Thank you to everyone who tried out the betas/release-candidates and/or gave feedback.

The v9 docs are now published to https://neutrinojs.org . For migration guide see: https://neutrinojs.org/migration-guide/

For any issues/feedback found from now on, please open as new GitHub issues :-)

**Comment by timkelty** [comment]
Thanks for all the work, @edmorley!

**Comment by brabeji** [comment]
Neutrino was and still is the best webpack config abstraction. Thank you @edmorley !
